### PR TITLE
[CUDA] upgrade cutlass to 3.5.0

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description
Upgrade cutlass to 3.5 to fix build errors using CUDA 12.4 or 12.5 in Windows
- [x] Upgrade cutlass to 3.5.0.
- [x] Fix flash attention build error with latest cutlass header files and APIs. This fix is provided by @wangyems.
- [x] Update efficient attention to use new cutlass fmha interface.
- [x] Patch cutlass to fix `hrsqrt` not found error for sm < 53.
- [x] Disable TF32 Staged Accumulation to fix blkq4_fp16_gemm_sm80_test build error for cuda 11.8 to 12.3.
- [x] Disable TRT 10 deprecate warnings. 

The following are not included in this PR:
* TRT provider replaces the deprecated APIs.
* Fix blkq4_fp16_gemm_sm80_test build error for cuda 12.4 or 12.5. This test is not built by default unless you add `--cmake_extra_defines  onnxruntime_ENABLE_CUDA_EP_INTERNAL_TESTS=ON` in build command.

To integrate to rel-1.18.1: Either bring in other changes (like onnx 1.16.1), or generate manifest and upload a new ONNX Runtime Build Time Deps artifact based on rel-1.18.1.

### Motivation and Context
https://github.com/microsoft/onnxruntime/issues/19891
https://github.com/microsoft/onnxruntime/issues/20924
https://github.com/microsoft/onnxruntime/issues/20953

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -46,6 +46,11 @@ else()
 set(CMAKE_CXX_STANDARD 17)
 endif()
 
+if (MSVC)
+  #  Make sure Visual Studio sets __cplusplus macro correctly: https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zc:__cplusplus")
```
Comment (snnn) [inline cmake/CMakeLists.txt:51]: Could you please explain why this change is needed for this PR? Is it only required by cutlass 3.5.0?

Comment (tianleiwu) [reply cmake/CMakeLists.txt:51]: See https://github.com/NVIDIA/cutlass/issues/1403.
Right now it only impacts cutlass build using nvcc from CUDA >= 12.4.

Comment (chilo-ms) [reply cmake/CMakeLists.txt:51]: Just FYI, https://github.com/microsoft/onnxruntime/issues/20923
We are working on a fix for TRT EP in response to adding `/Zc:__cplusplus`

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/threadblock/quantb_mma_multistage.h [hunk 2]
```diff
@@ -490,7 +490,10 @@ class QuantBMmaMultistage :
     // accuracy, where each mainloop iteration first accumulates into a temporary
     // set of freshly-cleared accumulators, which are subsequently added to the
     // final accumulator set.
-    static bool const kStagedAccumulation = arch::UseStagedAccumulation<typename Operator::MathOperator>::value;
+
+    // Change the following to avoid build error: class "cutlass::arch::OpMultiplyAdd" has no member "ElementA".
+    //     kStagedAccumulation = arch::detail::UseStagedAccumulation<typename Operator::MathOperator>::value;
+    static bool const kStagedAccumulation = false;
```
Comment (tianleiwu) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/threadblock/quantb_mma_multistage.h:496]: @chenfucn, please review this change. I think the staged accumulation is for TF32, but the q4gemm uses fp16 so it has no impact?

File: onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/fmha_launch_template.h [hunk 3]
```diff
@@ -75,12 +75,8 @@ struct RightPaddingBatchHook {
           batch_id * lse_dim * p.num_heads + head_id * lse_dim + query_start;
     }
 
-    // Custom masking
-    if (p.causal_diagonal_ptr) {
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/fmha_launch_template.h:79]: custom masking been removed?

Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/fmha_launch_template.h:79]: Right. It was removed in cutlass, and it is not used in our code base so I think it is fine.

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/threadblock/quantb_mma_multistage.h [hunk 4]
```diff
@@ -490,7 +490,10 @@ class QuantBMmaMultistage :
     // accuracy, where each mainloop iteration first accumulates into a temporary
     // set of freshly-cleared accumulators, which are subsequently added to the
     // final accumulator set.
-    static bool const kStagedAccumulation = arch::UseStagedAccumulation<typename Operator::MathOperator>::value;
+
```
Comment (wangyems) [inline onnxruntime/core/mickey/cutlass_ext/q4gemm/threadblock/quantb_mma_multistage.h:493]: arch::detail::...?
https://github.com/NVIDIA/cutlass/blob/637b15906358191cb4238af419d408a65819d7ec/include/cutlass/gemm/threadblock/mma_multistage.h#L160

**Review by snnn — CHANGES_REQUESTED** [review]
Do not disable C4996. See https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md

If the warning was generated when compiling an external *.cc/*.cpp file that is not part of our source tree, usually it wouldn't cause a build failure since we do not treat some warnings as errors
Otherwise, use tricks like https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/common/eigen_common_wrapper.h

File: onnxruntime/core/mickey/cutlass_ext/q4gemm/threadblock/quantb_mma_multistage.h (hunk 4 above)
Comment (tianleiwu) [reply onnxruntime/core/mickey/cutlass_ext/q4gemm/threadblock/quantb_mma_multistage.h:493]: I've tried such change, which has build error: class "cutlass::arch::OpMultiplyAdd" has no member "ElementA".

**Review by wangyems — APPROVED** [review]

**Review by snnn — APPROVED** [review]

**Review by faxu — APPROVED** [review]
